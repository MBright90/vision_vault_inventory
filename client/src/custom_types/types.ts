export interface genre_type {
    _id: string,
    __v: number,
    name: string,
    products: string[],
}

export interface product_type {
    _id: string,
    __v: number,
    name: string,
    description: string,
    price: number,
    number_in_stock: number,
    genres: Array<{
        name: string,
        _id: string
    }>,
    type: {
        name: string,
        _id: string
    },
    stock_last_updated: Date,
    last_updated: Date
}

export interface productType_type {
    _id: string,
    __v: number,
    products: string[],
    name: string
}
