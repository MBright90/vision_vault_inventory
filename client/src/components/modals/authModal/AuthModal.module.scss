@import '~@style/_variables.scss';

.modalBackground {
    background-color: rgba(100, 100, 100, 0.8);
    bottom: 0;
    display: grid;
    left: 0;
    position: absolute;
    place-content: center;
    top: 0;
    right: 0;
}

.modalContainer {
    align-items: center;
    background-color: $dark-red;
    color: $light-cream;
    display: flex;
    flex-direction: column;
    gap: 15px;
    padding: 30px;
    position: relative;

    .closeModal {
        background: none;
        color: $light-cream;
        height: 20px;
        padding: 3px !important;
        position: absolute;
        right: 5px;
        top: 5px;
        width: 20px !important;
    }

    input {
        width: 96%;
    }
    
    .buttonsContainer {
        display: flex;
        gap: 5%;
        justify-content: center;

        .confirmButton {
            margin-top: 15px;
            padding: 8px 30px;
            width: fit-content;
        }
    }
}