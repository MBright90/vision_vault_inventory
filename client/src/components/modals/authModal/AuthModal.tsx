import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faXmark } from "@fortawesome/free-solid-svg-icons";

import style from './AuthModal.module.scss';

interface AuthModalProps {
    closeModal: () => void,
    callback: (() => void) | (() => Promise<void>)
}

const AuthModal: React.FC<AuthModalProps> = ({ closeModal, callback }) => {
    return (
        <div className={style.modalBackground}>
            <div className={style.modalContainer}>
                <button className={style.closeModal} onClick={closeModal}><FontAwesomeIcon icon={faXmark}/></button>
                <p>Enter admin password to confirm:</p>
                <input className={style.passwordInput}/>
                <button className={style.confirmButton} onClick={() => { void callback(); closeModal(); }}>
                    Confirm
                </button>
            </div>
        </div>
    );
};

module.exports = AuthModal;