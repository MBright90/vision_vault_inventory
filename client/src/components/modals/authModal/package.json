{
    "name": "auth-modal",
    "version": "0.0.0",
    "main": "AuthModal.tsx",
    "dependencies": {
        "react": "^18.2.0"
    }
}
