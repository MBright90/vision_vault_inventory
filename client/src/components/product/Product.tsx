import React, { useState } from "react";

import capitalize from "@utilities/capitalize";
import style from "./Product.module.scss";
import type { product_type } from "@custom_types/types";

interface ProductProps {
    product: product_type
}

interface product_genre {
    name: string,
    _id: string
}

const Product: React.FC<ProductProps> = ({ product }) => {
    const [isEditingStock, setIsEditingStock] = useState<boolean>(false);
    const [editingStockValue, setEditingStockValue] = useState<number>(0);

    const productGenres: React.ReactNode[] = product.genres.map((genre: product_genre) => {
        return <span key={genre._id}>{capitalize(genre.name)}</span>;
    });

    const lastUpdatedArr: string[] = product.stock_last_updated.split('T');
    const lastUpdatedDate: string[] = lastUpdatedArr[0].split('-');
    const formattedLastUpdated: string = `${lastUpdatedArr[1].split('.')[0]} ${lastUpdatedDate[2]}/${lastUpdatedDate[1]}/${lastUpdatedDate[0]}`;

    // Stock update funcs
    const beginStockUpdate = (): void => {
        setIsEditingStock(true);
    };

    const changeStockEditValue = (e: React.ChangeEvent<HTMLInputElement>): void => {
        const newValue = parseInt(e.target.value);
        console.log(newValue);
        setEditingStockValue(newValue);
    };

    const updateStock = (): void => {
        console.log();
    };

    // Editing funcs
    const beginProductEdit = (): void => {
        console.log('beginning product edit');
    };

    // const editProduct = () => {

    // };

     return (
        <div className={style.productContainer}>
            <p className={style.productName}>{product.name}</p>
            <p className={style.productDetails}>{capitalize(product.type.name)} / {productGenres}</p>
            <p className={style.productDescription}>{product.description}</p>
            <div className={style.inStockContainer}>
                <button className={style.stockButton} onClick={beginStockUpdate}>Update stock</button>
                <p>In stock: { product.number_in_stock > 0 ? product.number_in_stock : 'NO' }</p>

                { isEditingStock ? <>
                    <input type='number' className={style.stockUpdate} value={editingStockValue} onChange={(e) => { changeStockEditValue(e); }} min={-100} max={100}/>
                    <button>Submit</button>
                    <button onClick={() => { setIsEditingStock(false); setEditingStockValue(0); }}>Cancel</button>
                </> : null }

            </div>
            <div className={style.editContainer}>
                <button className={style.editButton} onClick={beginProductEdit}>Edit details</button>
                <p>Last Edited: {formattedLastUpdated}</p>
            </div>
        </div>
     );
};

export default Product;

// name
// description
// price
// number_in_stock/stock_last_updated
// genres
// type
// last_updated
// edit button
// add stock button - admin
// sell one stock button
