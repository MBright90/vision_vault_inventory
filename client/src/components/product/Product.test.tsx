import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom"
import Product from "./Product";
import { MemoryRouter } from "react-router-dom";
import { product_type } from "@custom_types/types";
import { ReqOptions } from "@components/modals/authModal/AuthModal";

const mockProduct: product_type = {
    _id: 'test-id',
    __v: 0,
    name: 'test-name',
    description: 'test-description',
    price: 10,
    number_in_stock: 100,
    type: { _id: 'test-type-id', name: 'test-type-name' },
    genres: [
        { name: 'test-genre-one', _id: 'test-genre-id-1' },
        { name: 'test-genre-two', _id: 'test-genre-id-2' },
    ],
    stock_last_updated: '2024-06-05T09:09:10.297+00:00',
    last_updated: '2024-06-05T09:09:10.297+00:00',
}

jest.mock('@components/modals/confirmModal/ConfirmModal', () => ({
    __esModule: true,
    default: (props: { children: React.ReactNode[], closeModal: () => {} }) => {
        return (
            <div data-testid="mock-confirm-modal">
                {props.children}
            </div>
        )
    }
}));

jest.mock('@components/modals/authModal/AuthModal', () => ({
    __esModule: true,
    default: (props: { closeModal: () => {}, reqOptions: ReqOptions, productId: string, displayModalFunc: (message: string) => {} }) => {
        return (
            <div data-testid="mock-auth-modal">
                {props.reqOptions.body}
            </div>
        )
    }
}));

test('matches snapshot', () => {
    const { container } = render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
});

test('renders product information correctly', () => {
    render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    expect(screen.getByText(/^test-name/i)).toBeInTheDocument();
    expect(screen.getByText(/^test-type-name/i)).toBeInTheDocument();
    expect(screen.getByText(/^test-genre-one, test-genre-two/i)).toBeInTheDocument();
    expect(screen.getByText(/^test-description/i)).toBeInTheDocument();
    expect(screen.getByText(/^Â£10/i)).toBeInTheDocument();
    expect(screen.getByText(/^In stock: 100/i)).toBeInTheDocument();
});

test('renders update stock button', () => {
    render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    expect(screen.getByText(/^Update stock/i)).toBeInTheDocument();
});

test('displays number input and control buttons when updating stock', async () => {
    render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    const updateStockBtn = screen.getByText(/Update stock/i);
    fireEvent.click(updateStockBtn);

    await waitFor(() => {
        const updateStockInput = screen.getByRole('textbox', { name: 'update-stock' });
        expect(updateStockInput).toBeInTheDocument();
        expect(screen.getByText(/Submit/i)).toBeInTheDocument();
        expect(screen.getByText(/Cancel/i)).toBeInTheDocument();
    });
});

test('stock input value updates upon change', async () => {
    render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    const updateStockBtn = screen.getByText(/Update stock/i);
    fireEvent.click(updateStockBtn);

    await waitFor(() => {
        const updateStockInput = screen.getByRole('textbox', { name: 'update-stock' });
        fireEvent.change(updateStockInput, { target: { value: 20 } });
        expect(updateStockInput).toHaveValue(20);
    });
});

test('updates a positive stock change value', async () => {
    global.fetch = jest.fn();

    render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    const updateStockBtn = screen.getByText(/Update stock/i);
    fireEvent.click(updateStockBtn);

    await waitFor(() => {
        const updateStockInput = screen.getByRole('textbox', { name: 'update-stock' });
        fireEvent.change(updateStockInput, { target: { value: 20 } });

        const submitBtn = screen.getByText(/Submit/i);
        fireEvent.click(submitBtn);

        expect(screen.getByText(/In stock: 30/i)).toBeInTheDocument();
    });
});

test('updates a positive stock change value', async () => {
    global.fetch = jest.fn();

    render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    const updateStockBtn = screen.getByText(/Update stock/i);
    fireEvent.click(updateStockBtn);

    await waitFor(() => {
        const updateStockInput = screen.getByRole('textbox', { name: 'update-stock' });
        fireEvent.change(updateStockInput, { target: { value: -5 } });

        const submitBtn = screen.getByText(/Submit/i);
        fireEvent.click(submitBtn);

        expect(screen.getByText(/In stock: 5/i)).toBeInTheDocument();
    });
});

test('displays ConfirmModal with success message on stock update success', async () => {
    global.fetch = jest.fn((): Promise<Response> => 
        Promise.resolve({
            ok: true,
        } as Response)
    );

    render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    const updateStockBtn = screen.getByText(/Update stock/i);
    fireEvent.click(updateStockBtn);

    await waitFor(() => {
        const updateStockInput = screen.getByRole('textbox', { name: 'update-stock' });
        fireEvent.change(updateStockInput, { target: { value: 1 } });

        const submitBtn = screen.getByText(/Submit/i);
        fireEvent.click(submitBtn);

        expect(screen.getByTestId('mock-confirm-modal')).toBeInTheDocument();
        expect(screen.getByText(/^Stock updated/i)).toBeInTheDocument();
    });
});

test('displays ConfirmModal with error message on stock update error', async () => {
    global.fetch = jest.fn((): Promise<Response> => 
        Promise.resolve({
            ok: false,
        } as Response)
    );

    render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    const updateStockBtn = screen.getByText(/Update stock/i);
    fireEvent.click(updateStockBtn);

    await waitFor(() => {
        const updateStockInput = screen.getByRole('textbox', { name: 'update-stock' });
        fireEvent.change(updateStockInput, { target: { value: 1 } });

        const submitBtn = screen.getByText(/Submit/i);
        fireEvent.click(submitBtn);

        expect(screen.getByTestId('mock-confirm-modal')).toBeInTheDocument();
        expect(screen.getByText(/Error updating stock amount/i)).toBeInTheDocument();
    });
});

test('displays AuthModal when user requests product deletion', async () => {
    render(
        <MemoryRouter>
            <Product product={mockProduct} />
        </MemoryRouter>
    );

    const deleteBtn = screen.getByText(/^Delete/i);
    fireEvent.click(deleteBtn);

    await waitFor(() => {
        expect(screen.getByTestId('mock-auth-modal')).toBeInTheDocument();
    });
});