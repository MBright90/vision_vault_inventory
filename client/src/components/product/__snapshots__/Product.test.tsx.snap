// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div>
    <p>
      test-name
    </p>
    <p>
      Test-type-name
       
      <span>
        |
      </span>
       
      <span>
        Test-genre-one
        , 
      </span>
      <span>
        Test-genre-two
        , 
      </span>
    </p>
    <p>
      test-description
    </p>
    <p>
      Â£
      10
    </p>
    <div>
      <button>
        Update stock
      </button>
      <p>
        In stock: 
        100
      </p>
    </div>
    <div>
      <a
        href="/edit/test-id"
      >
        Edit details
      </a>
      <p>
        Last Edited: 
        09:09:10 05/06/2024
      </p>
    </div>
    <div>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;
